### ISSUE TITLE
NuGet package size concerns

### ISSUE THREAD

**Issue by 0xced** [Mongo2Go/Mongo2Go#133]
The [current NuGet package](https://www.nuget.org/packages/Mongo2Go/3.1.3) (version 3.1.3) weights 115 MB.

Commit a932934998c298f0ea1386cc7b9cfbf7fdcc6f10 added support for Linux/arm64. Creating a new NuGet package with MongoDB 5.1.1 and the database tools 100.5.1 now weights 146 MB. It would weight 180 MB without stripping binaries (introduced in 67dd1e9f4dcf9797ab941bed6b26d9bc1d6a2e09).

I think this starts to become problematic. 146 MB is a lot, even by todays standards. Also, 4 different platforms/architectures are provided by the package but only one will be usable. For example, the x64 ELF `mongod` binary will only run on Linux, it is a waste to download it on macOS and Windows.

I can think of two solutions to this problem.

1. Split the Mongo2Go package like SkiaSharp is doing with [SkiaSharp.NativeAssets.Linux][1], [SkiaSharp.NativeAssets.macOS][2] and [SkiaSharp.NativeAssets.Win32][3]. We could keep the core `Mongo2Go` package for the core functionality and provide 4 packages for the native assets (i.e. binary executables): `Mongo2Go.Native.Linux.arm64`,  `Mongo2Go.Native.Linux.x64`,  `Mongo2Go.Native.macOS.arm64` and `Mongo2Go.Native.Windows.x64`.

2. Do not provide the binary executable as part of the NuGet package but download them on demand instead.

My preference goes to the second option because it solves the package size issue and it even enables the consumer to choose the MongoDB version it wants at runtime. This would also untie the MongoDB version from the Mongo2Go version. The first option would still require to produce new NuGet package for new MongoDB versions whereas the second option completely solves this issue.

I have already started working on a branch ([split-downloader-cli_parts](https://github.com/0xced/Mongo2Go/tree/split-downloader-cli_parts)) to re-use the code that was written for the MongoDownloader tool so that it becomes usable directly from Mongo2Go. The idea is for the consumer to be able to specify which version of MongoDB it wants using the familiar [NuGet version range syntax](https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#version-ranges). I have already implemented the [best version match](https://github.com/0xced/Mongo2Go/blob/split-downloader-cli_parts/src/MongoDownloader/MongoDbDownloader.cs#L79) in my branch (leveraging the [NuGet.Versioning](https://www.nuget.org/packages/NuGet.Versioning) package).

[1]: https://www.nuget.org/packages/SkiaSharp.NativeAssets.Linux/
[2]: https://www.nuget.org/packages/SkiaSharp.NativeAssets.macOS/
[3]: https://www.nuget.org/packages/SkiaSharp.NativeAssets.Win32/

**Comment by JohannesHoppe** [comment]
# 👍 👍 

Yes, the second version sounds like a good plan. But here we will have massive problems with proxies. In my job I do workshops for companies and it is always hell to get a simple download via NPM.

If we go this way, we absolutely have to consider environment variables with HTTP_PROXY, HTTPS_PROXY, NO_PROXY, if possible the Windows proxy settings and to be on the safe side also proxy settings by code.

**Comment by JohannesHoppe** [comment]
But on the other hand, we would no longer have any trouble finding the right folder. That would make the whole complicated search logic obsolete. I am very excited.

**Comment by JohannesHoppe** [comment]
According to you code you are going to make a CLI. That looks good. Are you also planning a mode where you just call Mongo2Go as usual (without a CLI) and it just downloads the latest version without asking?

**Comment by 0xced** [comment]
> Yes, the second version sounds like a good plan. But here we will have massive problems with proxies. In my job I do workshops for companies and it is always hell to get a simple download via NPM.

I get you, I have spent waaaaaay too much time and energy resolving proxy issues on company networks. I even had to write tools to diagnose proxy issues! 😅

> If we go this way, we absolutely have to consider environment variables with HTTP_PROXY, HTTPS_PROXY, NO_PROXY, if possible the Windows proxy settings and to be on the safe side also proxy settings by code.

Absolutely ! This is already planned actually. A [default `HttpClient`](https://github.com/0xced/Mongo2Go/blob/split-downloader-cli_parts/src/MongoDownloader/Options.cs#L19) instance is provided but could be replaced by a custom one. Note that `HttpClient` observes the `HTTP_PROXY`, `HTTPS_PROXY`, `ALL_PROXY` and `NO_PROXY` environment variables and the system settings by default so this should be seldom needed.

**Comment by 0xced** [comment]
> According to you code you are going to make a CLI. That looks good. Are you also planning a mode where you just call Mongo2Go as usual (without a CLI) and it just downloads the latest version without asking?

Yes absolutely. I just split the original `MongoDownloader` project which was the CLI for automatically downloading the MongoDB binaries into `MongoDownloader` (core downloader classes) + `MongoDownloader.CLI` (command line interface to download the packages). I split into two different projects to make sure the CLI-specific parts (Spectre.Console) are kept in the CLI tool only.

My plan is to eventually move all the classes from `MongoDownloader` into `Mongo2Go` and automatically use the downloader from there, automatically handling where to download the binaries.

This will inevitably be a breaking change. I think we should choose a sensible default for where to download the executable binaries that the user could override. I was thinking `~/Library/Application Support/Mongo2Go` for macOS, `%LocalAppData%\Mongo2Go` on Windows and `~/.local/share/Mongo2Go` for Linux.

**Comment by JohannesHoppe** [comment]
> This will inevitably be a breaking change.

Version bump to v4 for sure!

One last v3 to address #131 and #132 and to use the current downloader for the first and last time! 😇

**Comment by asimmon** [comment]
I've read a lot about `runtime.json`, I summarized my findings in https://github.com/asimmon/ephemeral-mongo/issues/2 .
**TLDR**: it is not a viable option.

I still wanted to do something about it while supporting new MongoDB versions, that's why I created https://github.com/asimmon/ephemeral-mongo

**Comment by JohannesHoppe** [comment]
Time passed, and it looks like we now have a package size limit of about 250 MB. I'm pretty sure it was lower in the past, that's why I was never able to add ARM packages if I remember correctly.
https://learn.microsoft.com/en-us/nuget/nuget-org/publish-a-package#package-size-limits

But hey, it looks like we can add ARM support now too (for all the people who have a new Mac) and just ignore all the fancy package splitting tech. If anyone wants to add that, PRs are always welcome. ❤️

**Comment by teneko** [comment]
> But hey, it looks like we can add ARM support now too (for all the people who have a new Mac) and just ignore all the fancy package splitting tech. If anyone wants to add that, PRs are always welcome. ❤️
I do not understand what you mean by that, can you elaborate this, @JohannesHoppe?

I am not a big fan of prepacking mongodb binaries via NuGet. As consumer of such package, how can you verify that the content of such package is the one that was originally published by mongodb without downloading and looking in the NuGet package and comparing the hashes of the mongodb binaries by yourself? Because of this I would prefer the on-demand approach a lot more, this does not turn us into a binary data broker and someone just needs to look into the source code _how_ the binaries are getting downloaded not _what_ binaries has been prepacked. This approach would eliminate to keep the mongodb binaries up-to-date; it can get very messy to introduce a mongodb binary nuget package per architecture per platform.

When considering download from the internet, we must implement a user-friendly interception interface for enabling putting a proxy in front of it, to enable something like this, conditionally:

```
var handler = new HttpClientHandler()
{
    Proxy = new WebProxy("http://proxyserver:port"),
    UseProxy = true
};

var client = new HttpClient(handler);
```

where the user input is properly checked and such. I would introduce a simple config file/or msbuild variables that the user can drop into the project folder/write into the csproj file, since we only need access to it on build time and copy it over to the output directory to properly get accessed by mongo2go entry assembly when the download logic is done in the mongo2go entry assembly.

Just another thought of how you could download: you introduce a binary-download-specific mongodb nuget package with such capability to download desired architectures completely at build time via msbuild targets. So you can already see download problems at build time and not first at runtime of the mongo2go entry assembly.

TL;DR: I see significant merit in both proposed approaches—splitting the binaries into architecture-specific NuGet packages and implementing an on-demand binary download mechanism. However, I lean toward the on-demand approach because it resolves the package size issue, decouples MongoDB versioning from Mongo2Go releases, and mitigates trust concerns related to prepackaged binaries. For the on-demand strategy, we should also support proxy configurations via simple settings files or MSBuild variables, ensuring compatibility in restricted network environments. I acknowledge that there are multiple ways to approach binary download and resolution, and I find myself overwhelmed by the variety of options available. Further input to narrow down the strategy would be greatly appreciated.

Depending on which strategy we decide on I could write a proposal/concept of how to implement this.

**Comment by JohannesHoppe** [comment]
@teneko , you deleted your last message, but I still want to answer it. 🙂

I appreciate your detailed thoughts on **on-demand binary downloading** versus **prepackaged binaries** in NuGet. While I see the advantages of an on-demand approach, I have **strong concerns** about making this switch. Let me outline my reasoning:

### **1. Proxy Issues: Not Every Build Server Has Internet Access**  
Many enterprise environments operate with **strict network restrictions**. While **NuGet proxies allow package downloads**, this **does not guarantee** that the **build server has internet access** to download MongoDB binaries at runtime.  
- **Example issue**: A build pipeline that pulls Mongo2Go from an internal NuGet mirror, but is **not allowed** to access external URLs.  
- **Risk**: If Mongo2Go switches to an on-demand approach, **some users may be unable to run their tests at all** due to network restrictions.

### **2. Ensuring Reproducible Builds Over Time**  
One of my **biggest concerns** is **reproducibility**. I want Mongo2Go users to be able to:  
✅ Pull an old Mongo2Go version from NuGet **years later** and still get the **exact same binary** that was tested and approved at that time.  
✅ Run builds in **offline environments** without depending on **MongoDB’s website availability**.  

- **What happens in an on-demand model?**  
  - If MongoDB **removes old versions** from their servers, **older Mongo2Go versions break**.
  - If MongoDB **moves their download URLs**, **we have to update Mongo2Go**, even if the functionality itself didn’t change.

By **packaging the binaries** in NuGet, I **capture a snapshot of history**, ensuring that every Mongo2Go release is **fully self-contained** and **always works** in the future.

### **3. Trust Model & Security Considerations**  
I fully understand the trust concerns of **prepackaged binaries** in NuGet, so I’ve designed **a trust-minimized workflow** to ensure that the binaries I distribute are not tampered with.  

#### **How I ensure binary integrity today**:  
✅ **Step 1:** One person **runs MongoDownloader**, downloads MongoDB binaries, and commits them.  
✅ **Step 2:** **Before merging**, I run MongoDownloader **again** to verify that the binaries are identical.  
✅ **Step 3:** **GitHub Actions builds the NuGet package**, ensuring my **local machine does not modify the final package** (to prevent malware injection).  
✅ **Step 4:** The **NuGet package should be 100% reproducible**, meaning anyone can run the build and get the **same package contents**.  

This means you **don’t have to blindly trust me**—you can verify it yourself by:  
```sh
dotnet pack
```
and comparing the package contents.

---

### **Final Thoughts: Why I Prefer Prepackaging**
While I see the **technical appeal** of an on-demand approach, I believe **prepackaging binaries into NuGet** is the best option for Mongo2Go due to:  
✅ **No dependency on MongoDB’s website availability**  
✅ **No proxy-related issues in CI/CD environments**  
✅ **Guaranteed reproducibility for historical builds**  
✅ **A trust-minimized process that allows verification**  

That said, I **fully encourage** any attempts to verify my process! If you (or anyone else) can **reproduce the Mongo2Go package** exactly as published, that would help prove the **integrity of the current system**.

Would love to hear your thoughts! 🙂🙏

**Comment by JohannesHoppe** [comment]
A concrete step here in **Mongo2Go 4.2.0**: the bundled MongoDB binaries are now **stripped**, cutting the package by ~22%. Even with the newly added Linux arm64 binaries, the package is **smaller than 4.1.0** was.

This doesn't close the bigger question you raised about shipping one package per platform/runtime — that's still open for discussion — but it's a meaningful reduction in the meantime. Feedback welcome once 4.2.0 is out.

**Comment by JohannesHoppe** [comment]
Update for **Mongo2Go 5.0.0**: a heads-up that the single-package approach was kept, and with MongoDB 8.0.26 the package is now ~248 MB (the 8.x binaries are ~1.9× larger than 4.4.x). It stays under nuget.org's 250 MiB hard limit — but only just, so the concern you raised here is very much still alive. To make five platforms (incl. native macOS arm64) fit, 5.0.0 strips the binaries, ships only `mongod`/`mongoimport`/`mongoexport`, commits them gzip-compressed, and max-Deflate-recompresses the final package. Neither of your proposals — per-platform split (SkiaSharp-style) or download-on-demand — was adopted for 5.0.0, and they remain the real long-term fixes, especially the download-on-demand direction your `split-downloader-cli_parts` branch already prototypes. Keeping this open, but once again I was able to squeeze everything into that ZIP file... 😎
